Add Dynamo-native prefill/decode routing behind feature flag (#4566)
tenstorrent/tt-inference-server: dynamo_frontend/DEPLOY.md

`deploy.sh` brings up an etcd-backed Dynamo stack on the `dynamo-net` Docker
network — **etcd** (discovery), **cpp_server** (worker), and
**dynamo-frontend** (HTTP gateway) — plus **Prometheus** and **Grafana** from
`tt-media-server/monitoring/`. Optional prefill modes are:

> `--prefill-direct`, `--prefill-gateway`, and `--dynamo-native-routing`. Gateway

Reviewer: nit but
`--dynamo-native-routing` "native" kind of confused me, I believe the full "dynamo native router" would also include Dynamos KB Block manager, which we aren't doing. Maybe just call it dynamo-routing, or dynamo-kv-routing?

Author: Agreed, I renamed it to `--dynamo-routing`

| `--llm-device-backend <name>` | `pipeline_manager`                   | `LLM_DEVICE_BACKEND` env on the worker                                          |
| `--prefill-gateway`      | off                                       | Start PrefillGateway and route decode prefill requests through it               |
| `--prefill-gateway-image <img>` | `tt-prefill-gateway:dev`          | PrefillGateway image; the default local image is built automatically if missing |
| `--prefill-gateway-prefill-bind <host:port>` | `0.0.0.0:7200`        | ZMQ prefill ROUTER bind endpoint                                                |
| `--prefill-workers <count>` | `1`                                  | Managed prefill worker count for gateway and native Dynamo routing              |
| `--prefill-direct`       | off                                       | Start one managed prefill worker connected directly to decode, without PrefillGateway |

> | `--dynamo-native-routing` | off                                      | Experimental: register decode/prefill pools with Dynamo and let Dynamo route prefills |

Reviewer: ```suggestion
| `--dynamo-native-routing` | off                                      | Experimental: register decode/prefill pools with Dynamo and let routing decisions up to Dynamo |
```

Since dynamo routes to decode after prefill is done + soon enough it will route directly to decode on small prompts, might be misleading

| `--no-monitoring`        | off                                       | Skip Prometheus + Grafana                                                       |

Fixed (not flags): network `dynamo-net`, container names `etcd` / `tt-cpp-worker`
/ `dynamo-frontend`, monitoring container names `dynamo-prometheus` /
`dynamo-grafana` / `dynamo-process-exporter`, optional gateway container name
`prefill-gateway`, managed prefill worker name prefix `tt-cpp-prefill-worker`,
frontend host port `8080`,
`LLM_DEVICE_BACKEND=pipeline_manager`, `MODEL_NAME=tt-cpp-server`.
`PREFILL_DIRECT_SOCKET_PORT` defaults to `9000` when `--prefill-direct` is used.
`PREFILL_WORKER_COUNT` can set the same value as `--prefill-workers`.
`DYNAMO_NATIVE_NAMESPACE` defaults to `dynamo` when
`--dynamo-native-routing` is used, producing the documented Dynamo endpoints
`dynamo.decode.generate` and `dynamo.prefill.generate`.

`HF_TOKEN` (for gated models) and perf knobs (`ROUTER_MODE`, `DYN_TOKENIZER`,
`RAYON_NUM_THREADS`, `DYN_RUNTIME_*`, `RUST_LOG`, `DYN_TX_TRACE`,
`DYN_ENABLE_ANTHROPIC_API`, and Dynamo conditional-disagg

> `DYN_ROUTER_CONDITIONAL_*` variables) are read from the calling shell's

Reviewer: ```suggestion
`DYN_ENABLE_ANTHROPIC_API`), and Dynamo conditional-disagg
(`DYN_ROUTER_CONDITIONAL_*` variables) are read from the calling shell's
```
nit

   cpp_server inter-server ZMQ transport. The decode worker keeps short prompts
   local according to
   `MAX_TOKENS_TO_PREFILL_ON_DECODE` (default `1000`) and offloads larger
   prompt deltas to the managed prefill worker. Dynamo still receives requests
   through the decode worker; it does not select among prefill workers in this
   mode.
6. **PrefillGateway (optional)** — with `--prefill-gateway`, starts the
   gateway on `dynamo-net`. The default ZMQ mode binds `0.0.0.0:7200` for
   prefills and exposes metrics on container port `9091`; the script starts
   `--prefill-workers` managed `LLM_MODE=prefill` workers that connect to that
   bind endpoint. The Dynamo-registered worker runs as `LLM_MODE=decode` with
   `USE_PREFILL_GATEWAY=1` and `MAX_TOKENS_TO_PREFILL_ON_DECODE=0`, so decode
   requests route prefill work through the gateway.
7. **Dynamo-native routing (experimental)** — with
   `--dynamo-native-routing`, starts the worker as `LLM_MODE=decode` with
   `DYNAMO_NATIVE_ROUTING=1` on `dynamo.decode.generate`, then starts
   `--prefill-workers` managed `LLM_MODE=prefill` workers with
   `DYNAMO_ENDPOINT_ENABLED=1`,

>    `DYNAMO_WORKER_TYPE=prefill`, `DYNAMO_MODEL_TYPE=Prefill`, and endpoint

Reviewer: `DYNAMO_MODEL_TYPE=Prefill`
Is this correct? afaik model_type is either ModelType.Chat or ModelType.Completions as documented here https://docs.nvidia.com/dynamo/latest/backends/custom-backend/python-workers-lower-level

Author: In the dynamo router docs, prefill workers are registered with `ModelType.Prefill`:
https://docs.nvidia.com/dynamo/v-0-9-1/components/router/router-guide

The newer docs show the preferred contract as `worker_type=WorkerType.Prefill` with `model_type=ModelType.Empty`, so we should switch to `ModelType.Empty` once the frontend version we use supports that contract:
https://docs.nvidia.com/dynamo/components/router/disaggregated-serving

   `dynamo.prefill.generate`. The explicit `Prefill` model type keeps the MDC
   compatible with the released `ai-dynamo` frontend, which rejects
   `Tokens+Empty` even when `worker_type=Prefill` is present. Dynamo's

>    integrated prefill router owns the local-vs-remote prefill decision; when a

Reviewer: ```suggestion
   integrated router owns the local-vs-remote prefill decision; when a
```

Bit confusing, it routes everything not just prefill
